serverless/serverless
# Native VPC support

Closes #6673
Refs #6674

## What did you implement:

Adds native vpc support via dead-simple `vpc: true` config.

**NOTE:** This PR also includes ways to change the CIDR setup as shown below.

## How did you implement it:

Add a new `package:compileVpc` lifecycle and implemented the plugin to hook into this. The plugin in turn will create the CloudFormation resources according to #6673.

## How can we verify it:

Deploy the following service with the `--verbose` flag to see the outputs (`serverless deploy --verbose`).

### Simple setup

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

vpc: true

custom:
  idx: 0

functions:
  hello:
    handler: handler.handler
```

### Extended setup

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

vpc:
  cidrBlock: 11.0.0.0/16
  subnets:
    private:
      cidrBlock: 11.0.1.0/24
    public:
      cidrBlock: 11.0.2.0/24

custom:
  idx: 0

functions:
  hello:
    handler: handler.handler
```

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO
